### PR TITLE
feat: Reorg updates validator state

### PR DESCRIPTION
**Pull request by hudem1**

### CODE REVIEW

File: src/Nethermind.Arbitrum/Execution/Stateless/StateReconstructor.cs [hunk 1]
```diff
@@ -462,6 +462,50 @@ private void OnPruningFinished(object? sender, PruningEventArgs args)
...
+                    _logger.Info($"ReorgTo: reorging to block older than valid candidate block: reorg target {header.Number} with hash {header.Hash} and candidate header {_validHeaderCandidate.Number} with hash {_validHeaderCandidate.Hash}");
+                _trieStore.Dereference(_validHeaderCandidate.StateRoot!);
+                _validHeaderCandidate = null;
+            }
+        }
+
+        PreparedTrimBeyond(header);
+    }
+
+    private void PreparedTrimBeyond(BlockHeader header)
```
Comment (wurdum) [inline src/Nethermind.Arbitrum/Execution/Stateless/StateReconstructor.cs:534]: Please move private methods to go after public

**Review by wurdum — APPROVED** [review]

File: src/Nethermind.Arbitrum/Execution/Stateless/StateReconstructor.cs [hunk 2]
```diff
@@ -69,7 +69,7 @@ private sealed record PruningGate(BlockHeader ValidHeader, TaskCompletionSource
     const long BytesToEvictFromMemDb = 100 * 1024; // 100 KiB
 
     /// <summary>FIFO queue of pinned state roots; oldest entries are evicted when the queue exceeds <see cref="_maxStateRootsInMem"/>.</summary>
-    private readonly ConcurrentQueue<Hash256> _preparedQueue = new();
+    private ConcurrentQueue<BlockHeader> _preparedQueue = new();
```
Comment (AnkushinDaniil) [inline src/Nethermind.Arbitrum/Execution/Stateless/StateReconstructor.cs]: Is it safe? Is it reassigned and read only under `_reconstructionLock`? If so, maybe demote it to `Queue<BlockHeader>` and clear-and-refill instead of reassigning

File: src/Nethermind.Arbitrum/Execution/ArbitrumExecutionEngine.cs [hunk 3]
```diff
@@ -547,7 +547,7 @@ public ResultWrapper<EmptyResponse> PrepareForRecord(PrepareForRecordParameters
             numOfBlocks--; // genesis block doesn't need preparation, so recording one less block
 
         long lastHeaderNum = headerNum + (long)numOfBlocks;
-        List<Hash256> referencedStateRoots = new List<Hash256>((int)numOfBlocks);
+        List<BlockHeader> referencedStateRoots = new((int)numOfBlocks);
```
Comment (AnkushinDaniil) [inline src/Nethermind.Arbitrum/Execution/ArbitrumExecutionEngine.cs]: Is switching `_preparedQueue` from `Hash256` to full `BlockHeader` just to compare `.Number` during trim? If so, it seems to be wasteful, is it possible to use a tuple `(long Number, Hash256 StateRoot)`?

File: src/Nethermind.Arbitrum/Execution/Stateless/StateReconstructor.cs (hunk 2 above)
Comment (AnkushinDaniil) [reply src/Nethermind.Arbitrum/Execution/Stateless/StateReconstructor.cs]: Also since it's mutable, maybe worth adding tests `_maxStateRootsInMem` boundaries? How should it behave?

File: src/Nethermind.Arbitrum.Test/Execution/Stateless/MarkValidTests.cs
Comment (AnkushinDaniil) [inline src/Nethermind.Arbitrum.Test/Execution/Stateless/MarkValidTests.cs:1]: Is it possible to add a test for concurrent pruning vs reorg?

**Review by AnkushinDaniil — APPROVED** [review]

File: src/Nethermind.Arbitrum/Execution/Stateless/StateReconstructor.cs (hunk 2 above)
Comment (damian-orzechowski) [reply src/Nethermind.Arbitrum/Execution/Stateless/StateReconstructor.cs]: It is also modified in `MaybeCap`, but can this be called in parallel? If not, then I agree with @AnkushinDaniil.

**Review by damian-orzechowski — APPROVED** [review]

File: src/Nethermind.Arbitrum/Execution/Stateless/StateReconstructor.cs (hunk 2 above)
Comment (hudem1) [reply src/Nethermind.Arbitrum/Execution/Stateless/StateReconstructor.cs]: Clear-and-refill implies having a list holding valid entries, and then we'd have to iterate over that list again to push to the queue. So this adds an additional iteration.
Or I could do like this:
```csharp
 int count = _preparedQueue.Count;
for (int i = 0; i < count; i++)
{
    BlockHeader h = _preparedQueue.Dequeue();
    if (h.Number > header.Number)
        invalidHeaders.Add(h);
    else
        _preparedQueue.Enqueue(h); // keep (<= reorg target)
}
```
But i feel like it's less readable than the current solution.
What do you think ?

Comment (hudem1) [reply src/Nethermind.Arbitrum/Execution/Stateless/StateReconstructor.cs]: i already have tests for `_maxStateRootsInMem`, the behaviour is the queue gets trimmed in `PreparedAddTrim`

File: src/Nethermind.Arbitrum/Execution/ArbitrumExecutionEngine.cs (hunk 3 above)
Comment (hudem1) [reply src/Nethermind.Arbitrum/Execution/ArbitrumExecutionEngine.cs]: yes i thought about that but i thought it was cleaner to hold a BlockHeader and its actually just a pointer to the block header cached into the block tree. And its 8bytes instead of that tuple of 16 bytes.

File: src/Nethermind.Arbitrum/Execution/Stateless/StateReconstructor.cs (hunk 2 above)
Comment (AnkushinDaniil) [reply src/Nethermind.Arbitrum/Execution/Stateless/StateReconstructor.cs]: The snippet that you posted is exactly what I meant. The readability trade-off is real, but I'd actually prefer it because it lets `_preparedQueue` stay `readonly`. Currently, the field has to be made non-readonly to support `_preparedQueue = validHeaders`, which means we've lost the one static guarantee that every access goes through `_reconstructionLock`
